### PR TITLE
Loan/reservation system review — fix all 26 findings + updater hardening (#205)

### PR DESCRIPTION
**Pull request by fabiodalez-dev**
## Summary

Full review of the loan/reservation system (documented in `docs/loan-system-review-2026-07.md`, 27 findings: 5 high, 13 medium, 9 low) with fixes for all 26 actionable findings, plus the updater hardening that came out of #205.

### Loan/reservation fixes (selection)
- **H1** `LoanRepository::close()` now guards state (`attivo=1 AND stato IN ('in_corso','in_ritardo')`) — re-closing a returned loan can no longer falsify history.
- **H2** `createReservation` validates dates (no past start, no inverted range, duration capped by the new `loans.max_loan_duration_days` setting) — closes a memory-exhaustion vector on unbounded ranges.
- **H3+H4** the `loan_returned` / `reservation_expired` / `copy_unavailable_user` (+ new `reservation_cancelled`) email templates are now seeded in all 4 locales (installer + migration + hardcoded defaults + it_IT last-resort fallback), and the cron entry points bootstrap I18n — cron emails no longer fail or fall back to Italian on non-Italian installs.
- **H5** the book-detail JS uses the locale-aware injected API path instead of hardcoded `/api/libro`.
- **M-series**: repair-returns reassign future loans off the lost copy; canonical libri-first lock order everywhere; overdue claim/revert guarded; lost "book available" notifications recovered by a sweep; new `LoanEligibility` gate (suspended user / expired card) on every creation *and approval* path; `DateHelper::today()` unification; `deleteCopy` FK pre-check; web users can cancel their own pending/scheduled requests.
- **L-series**: pickup deadline clamped to the loan window, queue promotion loops until dry, `origine='diretto'` tagged, legacy availability endpoint authenticated, ICS RFC5545 line folding, cross-process maintenance cooldown.

### Updater hardening (#205)
- **Preflight writability dry-run**: the update aborts BEFORE touching any file if the PHP process cannot write somewhere, listing the exact paths (no more half-updated installs).
- **Self-heal**: unwritable paths owned by the PHP user are chmod-repaired automatically.
- **CLI upgrade path**: `php scripts/manual-upgrade.php <release-zip>` runs the full upgrade as the file owner — the escape hatch when web PHP cannot get write access.

### Migration
`migrate_0.7.26.sql` — seeds the 4 email templates × 4 locales (INSERT IGNORE, never overwrites admin-customized rows), adds `loans.max_loan_duration_days`, normalizes the `loan_overdue_admin` subject placeholder, replaces the false "removed from your wishlist" sentence. Idempotent; covered by `tests/migration-0.7.26.unit.php` (11 behavioral checks).

### Verification
Adversarial review of every fix (4 independent verification passes); 20/20 unit suites green (incl. 52 loan edge cases + 10 preflight checks with real filesystem permissions); full E2E 135/135; reinstall regression (fresh install + real admin-UI upgrade 0.7.25 → 0.7.26 with the migration applied by the production Updater) passed.

## Summary by CodeRabbit

* **Nuove funzionalità**
  * Aggiunti nuovi flussi per annullare prestiti e prenotazioni dall’area utente.
  * Introdotte nuove notifiche email per prestiti restituiti, prenotazioni scadute e annullate.
  * Migliorata la gestione multilingua dei template email.

* **Miglioramenti**
  * Le date e le scadenze ora risultano più coerenti con il fuso orario dell’app.
  * Aggiunti controlli più chiari su validità utente, durata massima delle richieste e disponibilità delle copie.

* **Bug Fixes**
  * Corrette diverse condizioni di errore e messaggi mostrati all’utente.
  * Reso più robusto l’aggiornamento e l’invio delle notifiche automatiche.

### CODE REVIEW

File: app/Controllers/LoanApprovalController.php [hunk 1]
```diff
@@ -234,14 +236,24 @@ public function approveLoan(Request $request, Response $response, mysqli $db): R
...
-            }
+            // DateHelper::today() incapsula già i fallback sul timezone (M9).
+            $today = DateHelper::today();
 
             $utenteId = (int) $loan['utente_id'];
+
+            // M7 — gate di idoneità anche in APPROVAZIONE: l'utente può essere
+            // stato sospeso (o la tessera può essere scaduta) tra la richiesta e
+            // l'approvazione admin. Stesso gate usato da store()/createReservation.
+            $eligibilityError = \App\Support\LoanEligibility::checkUser($db, $utenteId);
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/LoanApprovalController.php:255]: Fixed in 29fd2f93 — the user row is now locked FOR UPDATE before the eligibility gate, same pattern as PrestitiController::store().

File: app/Controllers/LoanApprovalController.php [hunk 2]
```diff
@@ -706,9 +724,37 @@ public function confirmPickup(Request $request, Response $response, mysqli $db):
...
+                ]));
+                return $response->withHeader('Content-Type', 'application/json')->withStatus(400);
+            }
+            $libroId = (int) $bookRow['libro_id'];
 
-            // Lock and verify loan is ready for pickup
+            // Lock della riga `libri` SENZA filtro deleted_at: come per le restituzioni
+            // (vedi LoanRepository::close), l'evasione di un prestito già approvato deve
+            // poter procedere anche se il libro è stato soft-deleted nel frattempo.
+            $lockBookStmt = $db->prepare("SELECT id FROM libri WHERE id = ? FOR UPDATE");
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/LoanApprovalController.php:760]: This is deliberate and documented in the code comment right above the lock: pickup/return/cancel flows must keep working when the book was soft-deleted after the loan was created (a librarian must be able to take back or cancel a loan of a deleted title). loan-edge-cases.unit.php has a regression test asserting exactly that ('soft-deleted book return closes loan & frees copy'). The forward flows (store/approve/renew/createReservation) all filter deleted_at IS NULL; only the backward flows carry the documented exception.

File: app/Controllers/LoanApprovalController.php [hunk 3]
```diff
@@ -1128,6 +1275,7 @@ public function cancelReservation(Request $request, Response $response, mysqli $
             }
         }
         $reservationId = (int) ($data['reservation_id'] ?? 0);
+        $reason = $data['reason'] ?? '';
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/LoanApprovalController.php]: Fixed in 29fd2f93 — reason is now scalar-guarded, trimmed and capped at 500 chars; HTML escaping already happens at the sink (EmailService::replaceVariables).

File: app/Controllers/ReservationManager.php [hunk 4]
```diff
@@ -66,6 +69,23 @@ public function flushDeferredNotifications(): void
...
+        $this->deferredExpiryNotifications = [];
+        if ($pendingExpiry !== []) {
+            $notificationService = new \App\Support\NotificationService($this->db);
+            foreach ($pendingExpiry as $expired) {
+                try {
+                    $notificationService->sendQueueReservationExpiredNotification($expired['email'], $expired['variables']);
+                } catch (\Throwable $e) {
+                    \App\Support\SecureLogger::error('Invio notifica scadenza prenotazione differita fallito', [
+                        'error' => $e->getMessage(),
+                    ]);
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/ReservationManager.php:91]: Fixed in 29fd2f93 — a false return of the deferred expiry send is now logged (SecureLogger::warning with a hashed recipient).

File: app/Controllers/ReservationManager.php [hunk 5]
```diff
@@ -520,15 +546,17 @@ private function sendReservationNotification(array $reservation): bool
...
+            // retryUnsentReservationNotifications() al run di manutenzione/cron
+            // successivo (M4) — prima nessuno le rileggeva e l'email era persa.
             if ($success) {
                 $stmt = $this->db->prepare("UPDATE prenotazioni SET notifica_inviata = 1 WHERE id = ?");
                 $stmt->bind_param('i', $reservation['id']);
                 $stmt->execute();
                 $stmt->close();
             } else {
-                error_log("ReservationManager: Email send failed for reservation ID {$reservation['id']}, will retry on next run");
+                error_log("ReservationManager: Email send failed for reservation ID {$reservation['id']}, will be retried by retryUnsentReservationNotifications() on next maintenance run");
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/ReservationManager.php]: Fixed in 29fd2f93 — switched to SecureLogger::warning with the reservation id in context.

File: app/Controllers/ReservationManager.php [hunk 6]
```diff
@@ -626,43 +740,93 @@ public function isBookAvailableForImmediateLoan($bookId, ?string $startDate = nu
...
+     * Uses transaction to ensure atomic update and queue reordering; le
+     * notifiche partono solo DOPO il commit (differite via
+     * flushDeferredNotifications() quando la transazione è esterna).
      *
      * @return int Number of reservations cancelled
      */
     public function cancelExpiredReservations(): int
     {
         $ownTransaction = $this->beginTransactionIfNeeded();
+        $expiryNotifications = [];
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Controllers/ReservationManager.php:763]: Fixed in 29fd2f93 — cancelExpiredReservations now rethrows when the transaction is external, same pattern as processBookAvailability; the own-transaction path logs via SecureLogger.

File: app/Models/SettingsRepository.php [hunk 7]
```diff
@@ -34,16 +34,20 @@ public function ensureTables(): void
...
                 subject VARCHAR(255) NOT NULL,
                 body LONGTEXT NOT NULL,
                 description TEXT,
                 active TINYINT(1) DEFAULT 1,
                 created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
-                updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP
+                updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+                UNIQUE KEY name_locale (name, locale)
             ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
         ";
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Models/SettingsRepository.php:52]: Fixed in 29fd2f93 — migrate_0.7.26.sql now has a section 0 with information_schema-gated guards: adds the locale column, drops the legacy UNIQUE(name) and adds UNIQUE(name,locale). Covered by a legacy-schema scenario in tests/migration-0.7.26.unit.php (checks 11-13).

File: app/Support/IcsGenerator.php [hunk 8]
```diff
@@ -280,30 +280,102 @@ private function formatEvent(array $event): string
...
+     * @return \DateTimeZone Resolved timezone object
+     */
+    private function appTimezone(): \DateTimeZone
+    {
+        $tz = (string) ConfigStore::get('app.timezone', $this->timezone);
+        try {
+            return new \DateTimeZone($tz);
+        } catch (\Throwable $e) {
+            return new \DateTimeZone('Europe/Rome');
+        }
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Support/IcsGenerator.php:340]: Fixed in 29fd2f93 — an invalid app.timezone now falls back to the constructor timezone first, Europe/Rome only as last resort.

File: app/Support/Updater.php [hunk 9]
```diff
@@ -2828,6 +2843,137 @@ private function removeOrphansInDirectory(string $currentDir, string $newDir, st
...
+            if ($item->isDir()) {
+                if (!is_dir($targetPath)) {
+                    $ancestor = $nearestExistingAncestor($targetPath);
+                    if (!$checkDirWritable($ancestor)) {
+                        $failures[$relativePath] = true;
+                    }
+                }
+            } elseif (file_exists($targetPath)) {
+                if (!is_writable($targetPath) && !$tryRepair($targetPath)) {
+                    $failures[$relativePath] = true;
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Support/Updater.php:2970]: Fixed in 29fd2f93 — a destination FILE that is a symlink is now unlinked and replaced with a real file in copyDirectory (no write-through), and the preflight treats it as replaceable (parent-writability check). Directory symlinks stay allowed on purpose: storage/uploads mounted elsewhere are legitimate setups. Behavioral test: checks 11-12 in tests/updater-preflight-writable.unit.php (outside target verified untouched).

File: app/Views/user_dashboard/prenotazioni.php [hunk 10]
```diff
@@ -446,6 +469,45 @@ function resolveCoverUrl(array $item, string $key = 'copertina_url'): string {
...
+    <div class="alert-outcome alert-outcome--success" role="status">
+      <i class="fas fa-check-circle" aria-hidden="true"></i>
+      <span><?= HtmlHelper::e($outcomeSuccess); ?></span>
+    </div>
+  <?php elseif ($outcomeError !== null): ?>
+    <div class="alert-outcome alert-outcome--error" role="alert">
+      <i class="fas fa-exclamation-circle" aria-hidden="true"></i>
+      <span><?= HtmlHelper::e($outcomeError); ?></span>
+    </div>
+  <?php endif; ?>
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Views/user_dashboard/prenotazioni.php:509]: Fixed in 29fd2f93 — all 4 new sites use htmlspecialchars(..., ENT_QUOTES, 'UTF-8'); the remaining HtmlHelper::e calls in this file are pre-existing code outside this PR's scope.

File: app/Views/user_dashboard/prenotazioni.php [hunk 11]
```diff
@@ -508,6 +570,16 @@ function resolveCoverUrl(array $item, string $key = 'copertina_url'): string {
...
+              <form method="post" action="<?= htmlspecialchars(url('/loan/cancel'), ENT_QUOTES, 'UTF-8') ?>"
+                    data-swal-confirm="<?= htmlspecialchars(__('Annullare questa richiesta di prestito?'), ENT_QUOTES, 'UTF-8') ?>"
+                    data-swal-confirm-button="<?= htmlspecialchars(__('Annulla richiesta'), ENT_QUOTES, 'UTF-8') ?>">
+                <input type="hidden" name="csrf_token" value="<?= HtmlHelper::e($csrfToken); ?>">
+                <input type="hidden" name="loan_id" value="<?= (int)$request['id']; ?>">
+                <button type="submit" class="btn-cancel">
+                  <i class="fas fa-trash" aria-hidden="true"></i>
+                  <span><?= __("Annulla richiesta") ?></span>
+                </button>
+              </form>
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Views/user_dashboard/prenotazioni.php:582]: Fixed in 29fd2f93 (see the reply on line 502's comment).

File: app/Views/user_dashboard/prenotazioni.php [hunk 12]
```diff
@@ -578,6 +650,20 @@ function resolveCoverUrl(array $item, string $key = 'copertina_url'): string {
...
+                    data-swal-confirm="<?= htmlspecialchars(__('Annullare questo prestito programmato?'), ENT_QUOTES, 'UTF-8') ?>"
+                    data-swal-confirm-button="<?= htmlspecialchars(__('Annulla prestito'), ENT_QUOTES, 'UTF-8') ?>">
+                <input type="hidden" name="csrf_token" value="<?= HtmlHelper::e($csrfToken); ?>">
+                <input type="hidden" name="loan_id" value="<?= (int)$loan['id']; ?>">
+                <button type="submit" class="btn-cancel">
+                  <i class="fas fa-trash" aria-hidden="true"></i>
+                  <span><?= __("Annulla prestito") ?></span>
+                </button>
+              </form>
+              <?php endif; ?>
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply app/Views/user_dashboard/prenotazioni.php:666]: Fixed in 29fd2f93 (see the reply on line 502's comment).

File: tests/migration-0.7.26.unit.php [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+$env = migLoadEnv($root . '/.env');
+$db = new mysqli(
+    null,
+    $env['DB_USER'] ?? '',
+    $env['DB_PASS'] ?? ($env['DB_PASSWORD'] ?? ''),
+    $env['DB_NAME'] ?? '',
+    0,
+    '/opt/homebrew/var/mysql/mysql.sock'
+);
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/migration-0.7.26.unit.php]: Fixed in 29fd2f93 — the socket is now configurable via E2E_DB_SOCKET or .env DB_SOCKET with a TCP fallback on DB_HOST/DB_PORT, and the test SKIPs cleanly (exit 0) when no database is reachable.

File: tests/updater-preflight-writable.unit.php [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+// 2 — il caso #205: parent dir non scrivibile, la NUOVA dir viene segnalata
+chmod("$dst/public/assets", 0555);
+$fails = $run();
+check(in_array('public/assets/swagger-ui', $fails, true),
+    '02 issue #205: new dir under unwritable parent is reported');
+
+// 3 — il file dentro la nuova dir non genera un secondo falso positivo separato dal padre
+check(count(array_filter($fails, fn($p) => str_starts_with($p, 'public/assets'))) <= 2,
+    '03 failures are collapsed to the offending dir(s), not one per file');
```
[the remark this answers is not in the document]
Comment (fabiodalez-dev) [reply tests/updater-preflight-writable.unit.php]: Fixed in 29fd2f93 — the assertion now explicitly checks the child file is not reported separately from the offending directory.
